test/tests/integrationTests.js

```diff
@@ -0,0 +1,322 @@
+
+		testItems = [];
+		for (let i = 0; i < 5; i++) {
+			testItems.push(yield createDataObject('item', {libraryID: Zotero.Libraries.userLibraryID}));
+		}
+		setAddEditItems(testItems[0]);
+		
+		sinon.stub(Zotero.Integration, 'getApplication', function(agent, command, docID) {
+			if (!applications[docID]) {
+				applications[docID] = new DocumentPluginDummy.Application();
+			}
+			return applications[docID];
+		});
+		
+		displayDialogStub = sinon.stub(Zotero.Integration, 'displayDialog', function(doc, dialogName, prefs, io) {
+			var ioResult = dialogResults[dialogName.substring(dialogName.lastIndexOf('/')+1, dialogName.length-4)];
+			if (typeof ioResult == 'function') {
+				ioResult = ioResult(doc, dialogName);
+			}
+			Object.assign(io, ioResult);
+			return Zotero.Promise.resolve();
+		});
+		
+		addEditCitationSpy = sinon.spy(Zotero.Integration.Document.prototype, 'addEditCitation');
+	});
+	
+	after(function() {
+		Zotero.Integration.getApplication.restore();
+		displayDialogStub.restore();
+		addEditCitationSpy.restore();
+	});
+	
+	describe('Document', function() {
+		describe('#addEditCitation', function() {
+			var setDocumentDataSpy;
+			var docID = this.fullTitle();
+			
+			before(function() {
+				setDocumentDataSpy = sinon.spy(DocumentPluginDummy.Document.prototype, 'setDocumentData');
+			});
+			
+			afterEach(function() {
+				setDocumentDataSpy.reset();
+			});
+			
+			after(function() {
+				setDocumentDataSpy.restore();
+			});
+			
+			it('should call doc.setDocumentData on a fresh document', function* () {
+				yield execCommand('addEditCitation', docID);
+				assert.isTrue(setDocumentDataSpy.calledOnce);
+			});
+			
+			it('should not call doc.setDocumentData on subsequent invocations', function* () {
+				yield execCommand('addEditCitation', docID);
+				assert.isFalse(setDocumentDataSpy.called);
+			});
+			
+			it('should not call doc.setDocumentData when document communicates for first time since restart, but has data', function* () {
+				Zotero.Integration.sessions = {};
+				yield execCommand('addEditCitation', docID);
+				assert.isFalse(setDocumentDataSpy.called);
+			});
+			
+			describe('when style used in the document does not exist', function() {
+				var docID = this.fullTitle();
+				var displayAlertStub;
+				var style;
+				before(function* () {
+					displayAlertStub = sinon.stub(DocumentPluginDummy.Document.prototype, 'displayAlert').returns(0);
+				});
+				
+				beforeEach(function() {
+					// 🦉birds?
+					style = {styleID: "http://www.example.com/csl/waterbirds", locale: 'en-US'};
+					
+					// Make sure style not in library
+					try {
+						Zotero.Styles.get(style.styleID).remove();
+					} catch (e) {}
+					initDoc(docID, {style});
+					displayDialogStub.reset();
+					displayAlertStub.reset();
+				});
+				
+				after(function* () {
+					displayAlertStub.restore();
+				});
+			
+				describe('when the style is not from zotero.org or citationstyles.org', function() {
```

Comment:
This isn't actually testing citationstyles.org, so shouldn't include that in the test description.